File: rtltcp.go
package nrsc5

import (
	"encoding/binary"
	"fmt"
	"net"
	"time"
)

// Pure-Go rtl_tcp client, ported from rtltcp.c.

// Tuner types (from librtlsdr rtl-sdr.h).
const (
	tunerE4000  = 1
	tunerFC0012 = 2
	tunerFC0013 = 3
	tunerFC2580 = 4
	tunerR820T  = 5
	tunerR828D  = 6
)

// rtl_tcp command opcodes.
const (
	rtltcpSetCenterFreq     = 0x01
	rtltcpSetSampleRate     = 0x02
	rtltcpSetTunerGainMode  = 0x03
	rtltcpSetTunerGain      = 0x04
	rtltcpSetFreqCorrection = 0x05
	rtltcpSetDirectSampling = 0x09
	rtltcpSetOffsetTuning   = 0x0a
	rtltcpSetBiasTee        = 0x0e
	rtltcpDongleID          = 0x0f
	rtltcpSetAGCMode        = 0x08
)

type rtltcp struct {
	conn      net.Conn
	tunerType uint32
	gainCount uint32
}

// dialRTLTCP connects to an rtl_tcp server and reads the dongle info.
func dialRTLTCP(addr string, timeout time.Duration) (*rtltcp, error) {
	conn, err := net.DialTimeout("tcp", addr, timeout)
	if err != nil {
		return nil, err
	}
	st := &rtltcp{conn: conn}

	// dongle_info_t: magic[4] tuner_type tuner_gain_count (network order)
	var info [12]byte
	if err := st.readFull(info[:]); err != nil {
		conn.Close()
		return nil, fmt.Errorf("rtl_tcp: reading dongle info: %w", err)
	}
	if string(info[:4]) != "RTL0" {
		conn.Close()
		return nil, fmt.Errorf("rtl_tcp: bad magic %q", info[:4])
	}
	st.tunerType = binary.BigEndian.Uint32(info[4:8])
	st.gainCount = binary.BigEndian.Uint32(info[8:12])
	return st, nil
}

func (st *rtltcp) readFull(buf []byte) error {
	off := 0
	for off < len(buf) {
		n, err := st.conn.Read(buf[off:])
		if err != nil {
			return err
		}
		off += n
	}
	return nil
}

// sendCmd writes a 5-byte command: 1-byte opcode + 4-byte big-endian param.
func (st *rtltcp) sendCmd(opc byte, param uint32) error {
	var cmd [5]byte
	cmd[0] = opc
	binary.BigEndian.PutUint32(cmd[1:], param)
	_, err := st.conn.Write(cmd[:])
	return err
}

func (st *rtltcp) setCenterFreq(freq uint32) error { return st.sendCmd(rtltcpSetCenterFreq, freq) }
func (st *rtltcp) setSampleRate(rate uint32) error { return st.sendCmd(rtltcpSetSampleRate, rate) }
func (st *rtltcp) setTunerGainMode(mode uint32) error {
	return st.sendCmd(rtltcpSetTunerGainMode, mode)
}
func (st *rtltcp) setTunerGain(gain uint32) error { return st.sendCmd(rtltcpSetTunerGain, gain) }
func (st *rtltcp) setFreqCorrection(ppm uint32) error {
	return st.sendCmd(rtltcpSetFreqCorrection, ppm)
}
func (st *rtltcp) setDirectSampling(on uint32) error {
	return st.sendCmd(rtltcpSetDirectSampling, on)
}
func (st *rtltcp) setOffsetTuning(on uint32) error { return st.sendCmd(rtltcpSetOffsetTuning, on) }
func (st *rtltcp) setBiasTee(on uint32) error      { return st.sendCmd(rtltcpSetBiasTee, on) }

// read samples into buf; returns bytes read (0 on EOF).
func (st *rtltcp) read(buf []byte) (int, error) {
	return st.conn.Read(buf)
}

// getTunerGains returns the supported gain values in tenths of a dB,
// from https://github.com/steve-m/librtlsdr/blob/master/src/librtlsdr.c
func (st *rtltcp) getTunerGains() []int {
	e4kGains := []int{-10, 15, 40, 65, 90, 115, 140, 165, 190, 215,
		240, 290, 340, 420}
	fc0012Gains := []int{-99, -40, 71, 179, 192}
	fc0013Gains := []int{-99, -73, -65, -63, -60, -58, -54, 58, 61,
		63, 65, 67, 68, 70, 71, 179, 181, 182,
		184, 186, 188, 191, 197}
	r82xxGains := []int{0, 9, 14, 27, 37, 77, 87, 125, 144, 157,
		166, 197, 207, 229, 254, 280, 297, 328,
		338, 364, 372, 386, 402, 421, 434, 439,
		445, 480, 496}

	switch st.tunerType {
	case tunerE4000:
		return e4kGains
	case tunerFC0012:
		return fc0012Gains
	case tunerFC0013:
		return fc0013Gains
	case tunerFC2580:
		return nil
	case tunerR820T, tunerR828D:
		return r82xxGains
	default:
		return nil
	}
}

// resetBuffer drains pending data, then reads cnt bytes to realign
// the stream after a gain change (from rtltcp.c).
func (st *rtltcp) resetBuffer(cnt int) error {
	st.conn.SetReadDeadline(time.Now().Add(50 * time.Millisecond))
	buf := make([]byte, 1024)
	recvd := 0
	for {
		n, err := st.conn.Read(buf)
		if err != nil {
			break
		}
		recvd += n
	}
	st.conn.SetReadDeadline(time.Time{})
	// if we read an odd number of bytes, read one more
	if recvd&1 != 0 {
		st.conn.SetReadDeadline(time.Now().Add(time.Second))
		st.conn.Read(buf[:1])
		st.conn.SetReadDeadline(time.Time{})
	}
	// then read cnt bytes
	remaining := cnt
	for remaining > 0 {
		toRead := len(buf)
		if remaining < toRead {
			toRead = remaining
		}
		st.conn.SetReadDeadline(time.Now().Add(2 * time.Second))
		n, err := st.conn.Read(buf[:toRead])
		if err != nil || n <= 0 {
			return fmt.Errorf("rtl_tcp: reset read failed: %w", err)
		}
		remaining -= n
	}
	st.conn.SetReadDeadline(time.Time{})
	return nil
}

func (st *rtltcp) close() {
	st.conn.Close()
}

File: python/nrsc5py/rtltcp.py
"""rtl_tcp client (mirrors rtltcp.go).

Protocol: the server sends a 12-byte greeting (magic "RTL0", tuner type,
tuner gain count, big-endian). Commands are 5 bytes: one opcode byte
plus a big-endian uint32 parameter. Samples then stream as cu8.
"""

from __future__ import annotations

import socket
import struct

OP_SET_CENTER_FREQ = 0x01
OP_SET_SAMPLE_RATE = 0x02
OP_SET_TUNER_GAIN_MODE = 0x03
OP_SET_TUNER_GAIN = 0x04
OP_SET_FREQ_CORRECTION = 0x05
OP_SET_AGC_MODE = 0x08
OP_SET_DIRECT_SAMPLING = 0x09
OP_SET_OFFSET_TUNING = 0x0A
OP_SET_BIAS_TEE = 0x0E

TUNER_NAMES = {1: "E4000", 2: "FC0012", 3: "FC0013", 4: "FC2580",
               5: "R820T", 6: "R828D"}

# Per-tuner gain tables in tenths of a dB (librtlsdr).
GAIN_TABLES = {
    1: [-10, 15, 40, 65, 90, 115, 140, 165, 190, 215, 240, 290, 340, 420],
    2: [-99, -40, 71, 179, 192],
    3: [-99, -73, -65, -63, -60, -58, -54, 58, 61, 63, 65, 67, 68, 70, 71,
        179, 181, 182, 184, 186, 188, 191, 197],
    4: [],
    5: [0, 9, 14, 27, 37, 77, 87, 125, 144, 157, 166, 197, 207, 229, 254,
        280, 297, 328, 338, 364, 372, 386, 402, 421, 434, 439, 445, 480,
        496],
    6: [0, 9, 14, 27, 37, 77, 87, 125, 144, 157, 166, 197, 207, 229, 254,
        280, 297, 328, 338, 364, 372, 386, 402, 421, 434, 439, 445, 480,
        496],
}


class RtlTcp:
    def __init__(self, host: str, port: int, timeout: float = 5.0):
        self.sock = socket.create_connection((host, port), timeout=timeout)
        greeting = self._read_full(12)
        if greeting[:4] != b"RTL0":
            raise ValueError("bad rtl_tcp magic")
        self.tuner_type, self.gain_count = struct.unpack(">II", greeting[4:12])
        self.sock.settimeout(None)

    def _read_full(self, n: int) -> bytes:
        buf = b""
        while len(buf) < n:
            chunk = self.sock.recv(n - len(buf))
            if not chunk:
                raise ConnectionError("rtl_tcp closed")
            buf += chunk
        return buf

    def _cmd(self, op: int, param: int):
        self.sock.sendall(struct.pack(">BI", op, param))

    def set_center_freq(self, freq: int):
        self._cmd(OP_SET_CENTER_FREQ, freq)

    def set_sample_rate(self, rate: int):
        self._cmd(OP_SET_SAMPLE_RATE, rate)

    def set_tuner_gain_mode(self, manual: bool):
        self._cmd(OP_SET_TUNER_GAIN_MODE, 1 if manual else 0)

    def set_tuner_gain(self, gain_tenths: int):
        self._cmd(OP_SET_TUNER_GAIN, gain_tenths)

    def set_freq_correction(self, ppm: int):
        self._cmd(OP_SET_FREQ_CORRECTION, ppm)

    def set_agc_mode(self, on: bool):
        self._cmd(OP_SET_AGC_MODE, 1 if on else 0)

    def set_direct_sampling(self, mode: int):
        self._cmd(OP_SET_DIRECT_SAMPLING, mode)

    def set_offset_tuning(self, on: bool):
        self._cmd(OP_SET_OFFSET_TUNING, 1 if on else 0)

    def set_bias_tee(self, on: bool):
        self._cmd(OP_SET_BIAS_TEE, 1 if on else 0)

    def read(self, n: int) -> bytes:
        return self.sock.recv(n)

    def reset_buffer(self, cnt: int):
        """Drain pending samples, then read cnt bytes to realign."""
        self.sock.settimeout(0.05)
        recvd = 0
        try:
            while True:
                chunk = self.sock.recv(1024)
                if not chunk:
                    break
                recvd += len(chunk)
        except TimeoutError:
            pass
        self.sock.settimeout(2.0)
        if recvd & 1:
            self.sock.recv(1)
        remaining = cnt
        while remaining > 0:
            chunk = self.sock.recv(min(1024, remaining))
            if not chunk:
                raise ConnectionError("rtl_tcp closed during reset")
            remaining -= len(chunk)
        self.sock.settimeout(None)

    def get_tuner_gains(self):
        return GAIN_TABLES.get(self.tuner_type, [])

    def close(self):
        self.sock.close()
